File: JavaKotlinDers8InstaCloneFirebase/app/src/main/java/com/example/javakotlinders8instaclonefirebase/FeedActivity.java
package com.example.javakotlinders8instaclonefirebase;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.widget.Toast;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.CollectionReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.EventListener;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.FirebaseFirestoreException;
import com.google.firebase.firestore.Query;
import com.google.firebase.firestore.QuerySnapshot;

import java.util.ArrayList;
import java.util.Map;

/*RecyclerView ListView'dan daha verimli çalıştığı için onu kullanacağız.
 * internetten veri çekeceğimiz zaman firebase kullanacağımız zaman vs daha verimli olması için kullanılıyor
 * listView kullanılması önerilmiyor.*/

public class FeedActivity extends AppCompatActivity {
    FirebaseAuth auth;
    FirebaseUser user;
    FirebaseFirestore firebaseFirestore;
    ArrayList<String> userEmailFromDb;
    ArrayList<String> userResimFromDb;
    ArrayList<String> userBaslikFromDb;
    ArrayList<String> userYorumFromDb;
    RecyclerView recyclerView;
    FeedRecyclerAdapter feedRecyclerAdapter;

    @Override //menüyü bağlamamız için kullanıyoruz bu metodu
    public boolean onCreateOptionsMenu(Menu menu) {
        //inflater:bir xml dosyası oluşturduğumuzda eğer onu kodla bağlayacaksak kullanıyoruz.
        MenuInflater menuInflater = getMenuInflater();
        menuInflater.inflate(R.menu.insta_options_menu, menu);

        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == R.id.cikisYapItem) {
            auth.signOut();
            Intent intentToCikis = new Intent(FeedActivity.this, MainActivity.class);
            startActivity(intentToCikis);
            finish();
        } else if (item.getItemId() == R.id.uploadItem) {
            Intent intentToUpload = new Intent(FeedActivity.this, UploadActivity.class);
            startActivity(intentToUpload);
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_feed);
        tanimla();
        verileriCek();
        verileriGoster();

    }

    private void tanimla() {

        auth = FirebaseAuth.getInstance();
        user = auth.getCurrentUser();
        firebaseFirestore = FirebaseFirestore.getInstance();

        userBaslikFromDb = new ArrayList<>();
        userEmailFromDb = new ArrayList<>();
        userResimFromDb = new ArrayList<>();
        userYorumFromDb = new ArrayList<>();

        recyclerView = findViewById(R.id.recyclerView);

    }

    public void verileriCek() {

        CollectionReference collectionReference = firebaseFirestore.collection("Posts");

        //verileri filtrelemek istiyosak where kullanıyoruz.

        /*collectionReference.whereEqualTo("baslik","merve").addSnapshotListener()...
        dediğimizde basliği merve olan verileri çekiyoruz sadece.*/

        // Bizde verileri tarihe göre çekeceğiz Yani verileri dizeceğiz.
        //           Tarihe göre, azalarak çektik. En Yeni atılan en üstte çıkıyor.

        collectionReference.orderBy("tarih", Query.Direction.DESCENDING).addSnapshotListener(new EventListener<QuerySnapshot>() {
            @Override
            public void onEvent(@Nullable QuerySnapshot value, @Nullable FirebaseFirestoreException error) {
                if (error != null) {
                    Toast.makeText(getApplicationContext(), error.getLocalizedMessage().toString(), Toast.LENGTH_LONG).show();
                }
                if (value != null) {
                    for (DocumentSnapshot snapshot : value.getDocuments()) {
                        Map<String, Object> data = snapshot.getData();

                        String baslik = (String) data.get("baslik");
                        String resim = (String) data.get("resim");
                        String useremail = (String) data.get("useremail");
                        String yorum = (String) data.get("yorum");
                        //aldığımız verileri bi dizide tutacağız.

                        userBaslikFromDb.add(baslik);
                        userEmailFromDb.add(useremail);
                        userResimFromDb.add(resim);
                        userYorumFromDb.add(yorum);
                        //şimdi bunları recyclerViewda gösterelim

                        //diziye bişey eklendiğinde recyclerView'ı uyarmamız gerekiyor.
                        feedRecyclerAdapter.notifyDataSetChanged();
                    }
                }
            }
        });
    }

    public void verileriGoster() {

        recyclerView.setLayoutManager(new LinearLayoutManager(FeedActivity.this));
        feedRecyclerAdapter = new FeedRecyclerAdapter(userEmailFromDb, userBaslikFromDb, userYorumFromDb, userResimFromDb);
        recyclerView.setAdapter(feedRecyclerAdapter);

    }

}

File: KotlinDers10FireBasee/app/src/main/java/com/example/kotlinders10firebasee/FeedActivity.kt
package com.example.kotlinders10firebasee

import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.widget.Toast
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.firebase.Timestamp
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import kotlinx.android.synthetic.main.activity_feed.*

/**
 * listelerken her bir listelediğim veri için etrafına border çizmek istiyorum bunun için
 * drawable-new-drawable resource file-border_stroke-ok
 *
 */
class FeedActivity : AppCompatActivity() {

    private lateinit var auth:FirebaseAuth
    private lateinit var database:FirebaseFirestore

    var userEmailFromDb:ArrayList<String> =ArrayList()  //boş bir arraylist tanımladık
    var userCommentFromDb:ArrayList<String> = ArrayList()
    var userImageFromDb:ArrayList<String> = ArrayList()

    var adapter:FeedRecyclerAdapter?=null

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        val menuInflater=menuInflater
        menuInflater.inflate(R.menu.options_menu,menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId==R.id.add_post){
            val intent=Intent(applicationContext,UploadActivity::class.java)
            startActivity(intent)
        }else if (item.itemId==R.id.exit){
            auth.signOut()
            val intent=Intent(applicationContext,MainActivity::class.java)
            startActivity(intent)
            finish()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_feed)
        auth=FirebaseAuth.getInstance()
        database= FirebaseFirestore.getInstance()
        getDataFromDatabase()
        listele()

    }

    fun getDataFromDatabase(){
/*
        //belirli bir documentteki idyi vererek içindeki verileri çekebiliriz
        database.collection("Posts").document("djshjhjfhs").get().addOnCompleteListener {  }
        //filtreleme yapabiliriz
        database.collection("Posts").whereEqualTo("userEmail","merve@gmail.com").addSnapshotListener()
*/

        database.collection("Posts").orderBy("date",Query.Direction.DESCENDING).addSnapshotListener { value, error ->
            if (error!=null){
                Toast.makeText(applicationContext,error.localizedMessage.toString(),Toast.LENGTH_LONG).show()

            }else{
                if (value!=null){
                    if (!value.isEmpty){
                        userEmailFromDb.clear()
                        userCommentFromDb.clear()
                        userImageFromDb.clear()

                        val documents=value.documents
                        for (document in documents){
                            val comment=document.get("comment") as String
                            val userEmail=document.get("userEmail") as String
                            val downloadUrl=document.get("downloadUrl") as String
                            val timestamp=document.get("date") as Timestamp
                            val date=timestamp.toDate()

                            userEmailFromDb.add(userEmail)
                            userCommentFromDb.add(comment)
                            userImageFromDb.add(downloadUrl)

                            adapter!!.notifyDataSetChanged()

                            //listView'ın bu tarz yerlerde kullanılması pek önerilmiyor. O yüzden recyclerView kullanacağız

                        }
                    }
                }
            }
        }
    }

    fun listele(){
        var layoutManager=LinearLayoutManager(this)
        recyclerView.layoutManager=layoutManager

        adapter=FeedRecyclerAdapter(userEmailFromDb,userCommentFromDb,userImageFromDb)
        recyclerView.adapter=adapter

    }

}
